# 9.2

In [1]:
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
from IPython.display import display, Markdown, Latex

%matplotlib inline

In [2]:
data = []
file = open('Regression.csv', 'r')
for str in file:
    for c in str.split():
        data.append(float(c))
file.close()   

y = np.array([data[0]])
for i in range(1, len(data)):
    y = np.append(y, data[i] - data[i - 1])
    
beta_1 = data[0]
beta_2 = (data[-1] - data[0]) / len(data)
sigma = sum((data[i] - data[i - 1] - (data[-1] - data[0])/len(data))**2 for i in range(1, len(data))) / (len(data) - 1)

Сведем задачу к линейной модели. $X_i = \beta_1 + i\beta_2 + \varepsilon_0 + ... + \varepsilon_n, i = 0, 1,...,n$ - расстояние, которое проехал трамвай на $i$ секунд. Тогда

$Y_i = X_i - X_{i-1} = \beta_2 + \varepsilon_i, i = 1, 2,...,n$ - расстояние, которое проезжает трамвай за $i$-ую секунду.

$Y = \begin{pmatrix} Y_0 \\ Y_1 \\ ... \\ Y_n \end{pmatrix} = \begin{pmatrix} \beta_1 + \varepsilon_0 \\ \beta_2 + \varepsilon_1 \\ ... \\ \beta_2 + \varepsilon_n \end{pmatrix}  => l = \begin{pmatrix} \beta_1 \\ \beta_2 \\ ...  \\ \beta_2 \end{pmatrix}, \theta = \begin{pmatrix} \beta_1 \\ \beta_2 \end{pmatrix}, Z = \begin{pmatrix} 1 & 0 \\ 0 & 1 \\ ... & ... \\ 0 & 1 \end{pmatrix}$

Найдем оценку наименьших квадратов для $\theta$:

$\hat{\theta} = \begin{pmatrix} \hat{\beta_1} \\ \hat{\beta_2} \end{pmatrix} $ = $ (Z^TZ)^{-1}Z^TY = \begin{pmatrix}  \begin{pmatrix} 1 & 0 & ... & 0 \\ 0 & 1 & ... & 1 \end{pmatrix} & \begin{pmatrix} 1 & 0 \\ 0 & 1 \\ ... \\ 0 & 1 \end{pmatrix}  \end{pmatrix} ^{-1}\begin{pmatrix} 1 & 0 & ... & 0 \\ 0 & 1 & ... & 1 \end{pmatrix}\begin{pmatrix} Y_0 \\ Y_1 \\ ... \\ Y_n \end{pmatrix} = \begin{pmatrix} 1 & 0 \\ 0 & n \end{pmatrix} ^{-1}\begin{pmatrix} 1 & 0 & ... & 0 \\ 0 & 1 & ... & 1 \end{pmatrix}\begin{pmatrix} Y_0 \\ Y_1 \\ ... \\ Y_n \end{pmatrix} $ = $ \begin{pmatrix} 1 & 0 \\ 0 & \frac{1}{n} \end{pmatrix} \begin{pmatrix} 1 & 0 & ... & 0 \\ 0 & 1 & ... & 1 \end{pmatrix}\begin{pmatrix} Y_0 \\ Y_1 \\ ... \\ Y_n \end{pmatrix} $ = $ \begin{pmatrix} 1 & 0 & ... & 0 \\ 0 & \frac{1}{n} & ... & \frac{1}{n} \end{pmatrix}\begin{pmatrix} Y_0 \\ Y_1 \\ ... \\ Y_n \end{pmatrix} $ = $ \begin{pmatrix} Y_0 \\ \sum_{i $ = $ 1}^n{Y_i\frac{1}{n}} \end{pmatrix} = \begin{pmatrix} X_0 \\ \frac{1}{n}(X_n - X_0) \end{pmatrix}$

In [3]:
print(beta_1)
print(beta_2)

14.7978
14.0360369


$\hat{\theta} = \begin{pmatrix} \beta_1 \\ \beta_2 \end{pmatrix} = \begin{pmatrix} 14.7978 \\ 14.0360 \end{pmatrix}$

Найдем несмещенную оценку $\sigma^2$:

$\hat{\sigma^2} = \frac{1}{n + 1 - k}||Y - Z\hat{\theta}||^2 $ = $ \frac{1}{n-1}||\begin{pmatrix} X_0 \\ X_1 - X_0 \\ ... \\ X_n - X_{n-1} \end{pmatrix} - \begin{pmatrix} 1 & 0 \\ 0 & 1 \\ ... & ... \\ 0 & 1 \end{pmatrix} \begin{pmatrix} X_0 \\ \frac{1}{n}(X_n - X_0) \end{pmatrix}||^2 $ = $ \frac{1}{n-1}||\begin{pmatrix} X_0 \\ X_1 - X_0 \\ ... \\ X_n - X_{n-1} \end{pmatrix} - \begin{pmatrix} X_0 \\ \frac{1}{n}(X_n - X_0) \\ ... \\ \frac{1}{n}(X_n - X_0) \end{pmatrix}||^2 $ = $ \frac{1}{n-1}||\begin{pmatrix} 0 \\ X_1 - X_0 - \frac{1}{n}(X_n - X_0) \\ ... \\ X_n - X_{n-1} - \frac{1}{n}(X_n - X_0) \end{pmatrix}||^2 = \frac{1}{n - 1}\sum_{i = 1}^n{(X_i - X_{i-1} - \frac{X_n - X_0}{n})^2}$

Оценка дисперсии отсчета времени:

$\varepsilon_i \sim N(0, \sigma^2), \varepsilon_i^t = \frac{\varepsilon_i}{\beta_2} => \varepsilon_i^t \sim N(0, \frac{\sigma^2}{\beta_2})$

$\hat{\sigma_t^2} = \frac{\hat{\sigma^2}}{\hat{\beta_2}}$

In [4]:
print(sigma)
print(sigma/beta_2)

2.316148815405628
0.16501444331523724


$\hat{\sigma^2} = 2.316$ - несмещенная оценка $\sigma^2$

$\hat{\sigma_t^2} = 0.165$ - оценка $\sigma^2_t$